File: data/readable_format/cpp/file835.cpp
#include <iostream>
#include <vector>
#include <cassert>

void per_out2(const std::vector<std::vector<double>>& input, int n, std::vector<std::vector<double>>& output, int& m) {
    int jo = 0;

    for (int j = 0; j < n; ++j) {
        if (input[2][j] != 0.0) {
            if (jo >= m) {
                std::cerr << "Output array is too small for the filtered data." << std::endl;
                return;
            }
            for (int i = 0; i < 3; ++i) {
                output[i][jo] = input[i][j];
            }
            ++jo;
        }
    }

    if (jo < m) {
        std::cerr << "Output array is larger than non-zero columns. Only " << jo << " columns are filled." << std::endl;
    }
    m = jo; // Update m to reflect the actual number of non-zero columns processed
}

int main() {
    // Test 1: All elements in the third row are non-zero
    std::vector<std::vector<double>> input1 = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    int m1 = 3;
    std::vector<std::vector<double>> output1(3, std::vector<double>(m1, 0));
    per_out2(input1, 3, output1, m1);
    assert(m1 == 3);

    // Test 2: Some elements in the third row are zero
    std::vector<std::vector<double>> input2 = {{1, 2, 3}, {4, 5, 6}, {7, 0, 0}};
    int m2 = 3; // Initially allocate more space than needed
    std::vector<std::vector<double>> output2(3, std::vector<double>(m2, 0));
    per_out2(input2, 3, output2, m2);
    assert(m2 == 1);

    // Test 3: All elements in the third row are zero
    std::vector<std::vector<double>> input3 = {{1, 2, 3}, {4, 5, 6}, {0, 0, 0}};
    int m3 = 3; // Initially allocate more space than needed
    std::vector<std::vector<double>> output3(3, std::vector<double>(m3, 0));
    per_out2(input3, 3, output3, m3);
    assert(m3 == 0);

    std::cout << "All tests passed." << std::endl;

    return 0;
}

File: data/readable_format/fortran/file835.f90
program test_per_out2
    implicit none
    double precision, allocatable :: input(:,:), output(:,:)
    integer :: n, m

    ! Test 1: All elements in the third row are non-zero
    n = 3
    m = 3
    allocate(input(3,n))
    allocate(output(3,m))
    input = reshape([1d0, 2d0, 3d0, 4d0, 5d0, 6d0, 7d0, 8d0, 9d0], shape(input))
    call per_out2(input, n, output, m)
    call print_output(output, m)

    ! Reset arrays for next test
    deallocate(input)
    deallocate(output)

    ! Test 2: Some elements in the third row are zero
    n = 3
    m = 3 ! Initially allocate more space than needed
    allocate(input(3,n))
    allocate(output(3,m))
    input = reshape([1d0, 2d0, 3d0, 4d0, 5d0, 6d0, 7d0, 0d0, 0d0], shape(input))
    call per_out2(input, n, output, m)
    call print_output(output, m)

    ! Reset arrays for next test
    deallocate(input)
    deallocate(output)

    ! Test 3: All elements in the third row are zero
    n = 3
    m = 3 ! Initially allocate more space than needed
    allocate(input(3,n))
    allocate(output(3,m))
    input = reshape([1d0, 2d0, 3d0, 4d0, 5d0, 6d0, 0d0, 0d0, 0d0], shape(input))
    call per_out2(input, n, output, m)
    call print_output(output, m)

    ! Cleanup
    deallocate(input)
    deallocate(output)

contains

    subroutine per_out2(in,n,out,m)
        implicit none
        double precision, intent(in) :: in(3,n)
        double precision, intent(out) :: out(3,m)
        integer, intent(in) :: n, m
        integer :: jo, j, i
        jo=0

        do j=1,n
            if (in(3,j).ne.0.0) then
                jo=jo+1

                if (jo > m) then
                    print *, "Output array is too small for the filtered data."
                    return
                end if

                do i=1,3
                    out(i,jo)=in(i,j)
                end do
            end if
        end do

        if (jo < m) then
            print *, "Output array is larger than non-zero columns. Only ", jo, " columns are filled."
        end if
    end subroutine per_out2

    subroutine print_output(out, m)
        double precision, intent(in) :: out(3,*)
        integer, intent(in) :: m
        integer :: i, j

        if (m == 0) then
            print *, "No output (all elements were zero)."
            return
        end if

        do j=1,m
            write(*,*) (out(i,j), i=1,3)
        end do
        write(*,*) '---'
    end subroutine print_output

end program test_per_out2